### ISSUE TITLE
Add fader support for Dimmers

### ISSUE THREAD

**Issue by esev** [pywemo/pywemo#254]
WeMo dimmers support a fade transition time for transitioning between brightness levels. This is supported for the Bridge device in pywemo, but not for Dimmers. It would be nice to have this supported in pywemo for Dimmers also.

`SetBinaryState(BinaryState, Duration, EndAction, brightness, fader, UDN, CountdownEndTime, deviceCurrentTime)`

The `fader` parameter appears to be the key to making this happen. The fader parameter requires 5 colon separated values (ex: `300:-1:1:0:33`). Here's a starting point on what these parameters might mean (this is my best guess at the parameters thus far):

1. Fade time in seconds
2. `-1`=fader not active; otherwise this appears to be a timestamp for when the fade started. Setting it to `0` appears to use the current timestamp.
3. Still trying to work out this one. 
   - Setting it to `0` starts a timer but doesn't cause any fade. At the end of the timer the Dimmer turns off and the brightness level remains at its original state.
   - Setting this to `1` causes the Dimmer to turn on and fade from the current brightness to the target brightness.
4. Floating point value representing the number of brightness levels to change per second. This is a computed value, setting this value seems to have no effect.
5. Target brightness - fade to this brightness level. 0 turns the device off after fading.

There is a green light on the device that remains active throughout the fade transition. If the device turns off after the fade, then this green light also turns off. If the device is on after the fade, this green light remains on too. The green light can be turned off by setting parameter 2 to `-1`. This seems to cancel the fade, leaving the brightness at its current setting (halting the fade at the current transitional brightness if it was still in progress).

If the Device is off, the previous brightness will be restored before activating the fader. If the light was at 100% brightness previously it will turn on to that brightness and then fade to the target brightness. This likely isn't desirable. If the device is off, it's probably best to set the brightness to 1% before starting a fade.

**Comment by Spectre5** [comment]
This would be a nice touch.  I had seen this attribute as well but I hadn't taken the time to investigate it closer.  All of my other devices contain a signature like this:

`SetBinaryState(BinaryState, Duration, EndAction, UDN, CountdownEndTime, deviceCurrentTime)`

I've wondered on these as well what `Duration` and the times are for.  If `Duration` is to make it turn on for a certain amount of time and then turn off automatically after some preset minutes --- then I'd be interested in that too.

**Comment by Spectre5** [comment]
As for the fader, that 3rd item seems to be related to toggling a button and is otherwise to control whether it is enabled or not?

```js
            /**
             * Toggle fade out button
             */
            updateToggle: function() {
                var fader  = this.model.get("fader"),
                    faderToggle = parseInt(fader.split(':')[2], 10);
                if (faderToggle === 1) {
                    this.$('.btnTurn').addClass('toggleOn');
                    this.model.set("sleepTimerToggle", 1);
                } else {
                    this.model.set("sleepTimerToggle", 0);
                }
            },
```


```js
            /**
             * Logic for sleep fader timer
             */
            runSleepTimer: function() {
                var fader = this.model.get('fader'),
                    sleepFaderSplit = fader.split(':'),
                    countDownTime = parseInt(sleepFaderSplit[0], 10),
                    referenceTime = parseInt(sleepFaderSplit[1], 10),
                    isFaderEnabled = parseInt(sleepFaderSplit[2], 10),
                    brightnessDelta = parseFloat(sleepFaderSplit[3]),
                    startBrightnessVal = parseInt(sleepFaderSplit[4], 10),
                    currentTime = Math.round(new Date().getTime() / 1000.0),
                    $elem = this.$('*[data-bind="customTime"]'),
                    endTime = referenceTime + countDownTime ,
                    leftTime = endTime - currentTime,
                    self = this;

                if (countDownTime < 0) {
                    countDownTime = 300;
                }

                this.model.set('customTime', countDownTime);
                this.model.set("sleepTimerToggle", isFaderEnabled);
                ...<snip>...
```

```js
            /**
             * sets the sleep fader timer
             */
            setSleepTimer: function() {
                var timer = parseInt(this.model.get('customTime'), 10),
                    currEpoch = Math.round(new Date().getTime() / 1000),
                    isFaderEnabled = parseInt(this.model.get('sleepTimerToggle'), 10);

                if (this.countDownInterval) {
                    currEpoch = -1;
                }

                this.showInitiatingStatus();
                var faderVal;
                faderVal = timer + ':' + currEpoch + ':' + isFaderEnabled+ ':0:0';

                this.model.saveActionWithVal(['fader'], faderVal).done(this.successSaveAction).fail(this.failFaderAction);
            },
```

**Comment by esev** [comment]
Ah ha. Yes, a sleep timer describes the behavior of the third parameter well. When set to `0` it behaves like a sleep timer, turning off the light after the specified timer period. When set to `1` it behaves like a fader.

Figuring out the `Duration` parameter sounds like a good idea too. My old [F7C027](https://www.belkin.com/us/support-product?pid=01t80000002xFCbAAM) switch has `SetBinaryState(BinaryState)`. All my other switch devices have the same that you mentioned: `SetBinaryState(BinaryState, Duration, EndAction, UDN, CountdownEndTime, deviceCurrentTime)`.

**Comment by esev** [comment]
Looking at the `setSleepTimer` function you mentioned, I see that the brightness is hard-coded to `0`. That could explain why the green light never goes off when you fade to a non-zero value - they may not have intended for that to happen. I also don't see any way to fade-On in the app.
